### PR TITLE
✨ Parse list notation

### LINKED ISSUE #151: Parse list notation in path
### Description
Update `toPath` to parse list notation (see #139).

This is in 1.0.0 to avoid breaking changes (in the behavior of `toPath`) in the next version.

#148 has to be fixed before this, a list path step would look like this :
`[list, ['id1', 'id2', 'id6']]`

### PR DESCRIPTION
**Pull request by nlepage**
**Issue :** fix #151

### CODE REVIEW

File: packages/immutadot/src/path/toPath.js [hunk 1]
```diff
@@ -117,14 +118,28 @@ const sliceNotationParser = map(
...
+    return [[prop, beforeNewSegment], ...stringToPath(rest)]
+  },
 )
 
-const pathSegmentEndedByBracketParser = map(
-  regexp(/^([^.[]*?)(\[.*)$/),
-  ([beforeBracket, atBracket]) => [[prop, beforeBracket], ...stringToPath(atBracket)],
+const pathSegmentEndedByNewSegment = map(
+  regexp(/^([^.[{]*)\.?([[{]?.*)$/),
+  ([beforeNewSegment, rest]) => [[prop, beforeNewSegment], ...stringToPath(rest)],
```
Comment (hgwood) [inline packages/immutadot/src/path/toPath.js]: Nice simplification!

File: packages/immutadot/src/path/toPath.js [hunk 2]
```diff
@@ -117,14 +118,28 @@ const sliceNotationParser = map(
...
+      props.push(propMatch[2] === undefined ? propMatch[4] : propMatch[2])
+    return props.length === 1 ? [[prop, props[0]], ...stringToPath(rest)] : [[list, props], ...stringToPath(rest)]
+  },
+)
+
+const incompleteListNotationParser = map(
+  regexp(/^(\{[^.[{]*)(.*)$/),
+  ([beforeNewSegment, rest]) => {
+    return [[prop, beforeNewSegment], ...stringToPath(rest)]
+  },
```
Comment (hgwood) [inline packages/immutadot/src/path/toPath.js]: Maybe the lambda could be written in the expression form (ie without braces and `return`)?

File: packages/immutadot/src/path/toPath.js [hunk 3]
```diff
@@ -117,14 +118,28 @@ const sliceNotationParser = map(
...
+const listPropRegexp = /^,?((?!["'])([^,]*)|(["'])(.*?[^\\])\3)(.*)/
+const listNotationParser = map(
+  regexp(/^\{(((?!["'])[^,}]*|(["']).*?[^\\]\2)(,((?!["'])[^,}]*|(["']).*?[^\\]\6))*)\}\.?(.*)$/),
+  ([rawProps, , , , , , rest]) => {
+    const props = []
+    if (rawProps.startsWith(',')) props.push('')
+    for (let propMatch = rawProps.match(listPropRegexp); propMatch !== undefined; propMatch = propMatch[5] === '' ? undefined : propMatch[5].match(listPropRegexp))
+      props.push(propMatch[2] === undefined ? propMatch[4] : propMatch[2])
+    return props.length === 1 ? [[prop, props[0]], ...stringToPath(rest)] : [[list, props], ...stringToPath(rest)]
+  },
```
Comment (hgwood) [inline packages/immutadot/src/path/toPath.js]: This piece of code is a little overwhelming. My advice would be:
- use destructuring to name the elements of `propMatch`
- use functional programming to make meaningful transformation names emerge (what is this loop doing? a reduction? a mapping?)

**Comment by nlepage** [comment]
I deliberatly changed the behavior of incomplete notation for the list.
Instead of running to the end of the path like array and slice do, we make a prop including the opening curly until the next segment.
I think most of the time this will give a result closer to what the user meant, and highlight his mistake with the opening curly included in the prop.

@hgwood @frinyvonnick What do you think ? If you are OK, I will comment on #113 in order to refactor array and slice to make the same.

I also mutualized the later parsers in only one.
I had not anticipated the removal of `str => [[prop, str]]`, but code coverage reported that this code was dead, and all the tests were still working ! The perks of having 100% coverage !

File: packages/immutadot/src/path/toPath.js [hunk 4]
```diff
@@ -117,14 +118,28 @@ const sliceNotationParser = map(
   ([sliceStart, sliceEnd, rest]) => [[slice, [toSliceIndex(sliceStart), toSliceIndex(sliceEnd)]], ...stringToPath(rest)],
 )
 
-const pathSegmentEndedByDotParser = map(
-  regexp(/^([^.[]*?)\.(.*)$/),
-  ([beforeDot, afterDot]) => [[prop, beforeDot], ...stringToPath(afterDot)],
+const listPropRegexp = /^,?((?!["'])([^,]*)|(["'])(.*?[^\\])\3)(.*)/
+const listNotationParser = map(
+  regexp(/^\{(((?!["'])[^,}]*|(["']).*?[^\\]\2)(,((?!["'])[^,}]*|(["']).*?[^\\]\6))*)\}\.?(.*)$/),
```
Comment (hgwood) [inline packages/immutadot/src/path/toPath.js]: We might want to split the quoted case and the bare case like I did for the bracket notation. I feel it yielded simpler regexes. These ones are a bit hard to read.

**Comment by hgwood** [comment]
@nlepage *their mistake :)

That last parser is dead because `pathSegmentEndedByNewSegment` already matches everything. Again, nice simplification.

As for the incomplete cases, I'm fine both ways.

**Comment by nlepage** [comment]
@hgwood I've used a generator to extract list props, and `propMatch` elements are now named as you suggested, what do you think ?
Regarding the main regexp, I don't see how to split quoted and bare case... Do you have any pointers ?

**Comment by frinyvonnick** [comment]
@nlepage i agree for incomplete cases !

**Comment by hgwood** [comment]
@nlepage Great improvement to `listNotationParser`, I think it is a lot faster to read and understand now.

As for splitting the regex, I see now that you cannot really do that as you may have lists with both quoted and unquoted props. I'm still a bit worried by the length of the regex and also the duplication. I think I may have an idea to mitigate that, but I don't want to block this PR on it.

My idea is to go full parser combinators to have modular, composable, easy-to-understand parsers so you could describe lists like `between("{}", interspersedBy(",", maybeBetween(`''""`)))`. Might be an interesting refactoring to undertake, but again, not worth blocking on.

File: packages/immutadot/src/path/toPath.js [hunk 5]
```diff
@@ -108,14 +109,33 @@ const sliceNotationParser = map(
   ([sliceStart, sliceEnd, rest]) => [[slice, [toSliceIndex(sliceStart, 0), toSliceIndex(sliceEnd)]], ...stringToPath(rest)],
 )
 
-const pathSegmentEndedByDotParser = map(
-  regexp(/^([^.[]*?)\.(.*)$/),
-  ([beforeDot, afterDot]) => [[prop, beforeDot], ...stringToPath(afterDot)],
+const listPropRegexp = /^,?((?!["'])([^,]*)|(["'])(.*?[^\\])\3)(.*)/
+function* extractListProps(rawProps) {
+  if (rawProps.startsWith(',')) yield ''
```
Comment (frinyvonnick) [inline packages/immutadot/src/path/toPath.js]: This case isn't tested 😢

File: packages/immutadot/src/path/toPath.js [hunk 6]
```diff
@@ -108,14 +109,33 @@ const sliceNotationParser = map(
...
+    const [, , bareProp, , quotedProp, rest] = listPropRegexp.exec(remProps)
+    yield bareProp === undefined ? quotedProp : bareProp
+    remProps = rest
+  }
+}
+
+const listNotationParser = map(
+  regexp(/^\{(((?!["'])[^,}]*|(["']).*?[^\\]\2)(,((?!["'])[^,}]*|(["']).*?[^\\]\6))*)\}\.?(.*)$/),
+  ([rawProps, , , , , , rest]) => {
+    const props = [...extractListProps(rawProps)]
```
Comment (frinyvonnick) [inline packages/immutadot/src/path/toPath.js]: Smart!

**Comment by frinyvonnick** [comment]
Nice work 👍

**Pull request merged**